test/integration/targets/one_host/tasks/main.yml

```diff
@@ -0,0 +1,221 @@
+# test code for the one_host module
+
+
+# ENVIRONENT PREPARACTION
+
+- set_fact: test_number= 0
+
+- name: "test_{{test_number}}: copy fixtures to test host"
+  copy:
+    src: testhost/tmp/opennebula-fixtures.json.gz
+    dest: /tmp
+  when:
+    - opennebula_test_fixture
+    - opennebula_test_fixture_replay
+
+
+# SETUP INITIAL TESTING CONDITION
+
+- set_fact: test_number={{ test_number | int + 1 }}
+
+- name: "test_{{test_number}}: ensure the tests hosts are absent"
+  one_host:
+    name: "{{ item }}"
+    state: absent
+    endpoint: "{{ opennebula_endpoint }}"
+    session: "{{ opennebula_session }}"
+    validate_certs: false
+  environment:
+    PYONE_TEST_FIXTURE: "{{ opennebula_test_fixture }}"
+    PYONE_TEST_FIXTURE_FILE: /tmp/opennebula-fixtures.json.gz
+    PYONE_TEST_FIXTURE_REPLAY: "{{ opennebula_test_fixture_replay }}"
+    PYONE_TEST_FIXTURE_UNIT: "test_{{test_number}}_{{ item }}"
+  with_items: "{{opennebula_test.hosts}}"
+  register: result
+
+# NOT EXISTING HOSTS
+
+- set_fact: test_number={{ test_number | int + 1 }}
+
+- name: "test_{{test_number}}: attempt to enable a host that does not exists"
+  one_host:
+    name: badhost
+    state: enabled
+    endpoint: "{{ opennebula_endpoint }}"
+    session: "{{ opennebula_session }}"
+    validate_certs: false
+  environment:
+    PYONE_TEST_FIXTURE: "{{ opennebula_test_fixture }}"
+    PYONE_TEST_FIXTURE_FILE: /tmp/opennebula-fixtures.json.gz
+    PYONE_TEST_FIXTURE_REPLAY: "{{ opennebula_test_fixture_replay }}"
+    PYONE_TEST_FIXTURE_UNIT: "test_{{test_number}}"
+  ignore_errors: true
+  register: result
+
+- name: "assert test_{{test_number}} failed"
+  assert:
+    that:
+    - result is failed
+    - result.msg == 'invalid host state ERROR'
+
+# ---
+
+- set_fact: test_number={{ test_number | int + 1 }}
+
+- name: "test_{{test_number}}: delete an unexisting host"
+  one_host:
+    name: badhost
+    state: absent
+    endpoint: "{{ opennebula_endpoint }}"
+    session: "{{ opennebula_session }}"
+    validate_certs: false
+  environment:
+    PYONE_TEST_FIXTURE: "{{ opennebula_test_fixture }}"
+    PYONE_TEST_FIXTURE_FILE: /tmp/opennebula-fixtures.json.gz
+    PYONE_TEST_FIXTURE_REPLAY: "{{ opennebula_test_fixture_replay }}"
+    PYONE_TEST_FIXTURE_UNIT: "test_{{test_number}}"
+  register: result
+
+- name: "assert test_{{test_number}} worked"
+  assert:
+    that:
+    - result.changed
```

Comment:
If the host doesn't exist and we try to delete it, the module should return `changed: false`.

Reply:
It is slightly different.
The host is in fact created, the module fails because it does not reach the "ENABLED" state.
Normally this will happen because there is a connectivity problem, and the control node cannot reach the host.
If you run the playbook again, the hsot will still not be "enabled" and your play will continue to fail, util the host reaches the desired state "enabled".

Until you fix the connectivity problem, the one_host module will continue to fail.

On the other hand you can always delete a host even if it is in ERROR state.

Reply:
So if I create a host with state `present`, in which state is it created, because there is no state `present`? I guess it's created i state `enabled`. The thing is that this should be documented, as well as other states.

Looking at the code all states except `absent` actually create a new host. So it should be documented like that:
* `absent` deletes the host
* `enabled`, `disabled`, `offline` create a host (if doesn't exist) and set the state, or sets the sate on the current host.
* `present` if the host exists, doesn't change the state, if it doesn't exist creates the host and sets the state to `DEFAULT_STATE` (we could probably discuss what should be the default state, because I guess `enabled` is the default, which from my experience is bad, `disabled` is a much better initial state because it allows for testing of the host before allowing the scheduler to deploy new VMs on it).

Did I get this right?

Also when I think about it, `state` doesn't have to have a default if you ask me, because `present` state isn't natural for host state in OpenNebula.  Maybe we can just remove the `present` state and set `state` to be empty by default. That means that if you just want to make some changes to the host, like the `template` part, you only specify the `name` and the `template` parameter. It would make the docs a lot simpler and easier to understand. I've been using OpenNebula for a long time and it was confusing for me.

Reply:
I see where you are going and you see the same problem I did.

The present state is satisfied by hosts that are in Enabled, Disabled or Offline states.

the rationale behind it is becase you want to separate "provisioning" from "maintenance" playbooks.

When I provisiong our datacenter I create the objects: hosts, networks, etc. and Ensure that they are created OK.

In maintenance, for example: applying patches: you are going to disable the host, add OS patches, reboot, and re-enable. (grossly speaking).

If re-run provisioning playbook, you dont want it to change the state of the host to any as long as it is present in the platform.

You don't want your provisioning playbooks to chage the state of the host as long as it is present.

Reply:
With this point I suggest we take some field experience before deciding forward.

Are you planing on running a physical host cluster? that would be awesome.

Reply:
I understand the provisioning part, didn't think of that use case completely. Then `present` is fine as is, I would just document what present means when the host is created. I can see that in the future I would like to have parameter e.g. `initial_state` which could define the initial state of the host when it's being created. Having that parameter you wouldn't even have to document what will be the initial state, besides saying that the value of `initial_state` is followed.

We are running a physical host cluster of OpenNebula, and this was the next module we were thinking to work one because we need it for host maintenance which you described.